### PR TITLE
Localizations for Material

### PR DESCRIPTION
**Pull request by HansMuller**
Added Material widget localizations for nine locales (including english):
- ar Arabic
- it Italian
- pt Portugese
- es Spanish
- fr French
- zh Chinese
- en English
- de German
- ja Japanese
- ru Russian

The actual localizations were produced automatically with the Google Translator Toolkit (https://support.google.com/translatortoolkit). The translator toolkit works with Application Resource Bundle (".arb") files, see packages/flutter/lib/src/material/i18n.

The arb files were converted into a small Map definition with dev/tools/gen_localizations.dart, see packages/flutter/lib/src/material/i18n/localizations.dart.

The MaterialLocalizations class looks up resources in the generated localizations Map.

### CODE REVIEW

File: packages/flutter/lib/src/material/material_localizations.dart [hunk 1]
```diff
@@ -2,35 +2,53 @@
...
 import 'package:flutter/widgets.dart';
 
-/// Interface for localized resource values for the material widgets.
-///
-/// This class provides a default placeholder implementation that returns
-/// hard-coded American English values.
+import 'i18n/localizations.dart';
+
+/// Localized strings for the material widgets.
 class MaterialLocalizations {
```
Comment (Hixie) [inline packages/flutter/lib/src/material/material_localizations.dart]: I think our default localization object should be really really simple. I'd prefer if we added the other languages as a separate package.

File: packages/flutter/lib/src/material/material_localizations.dart [hunk 2]
```diff
@@ -2,35 +2,53 @@
...
-  /// which only provides American English strings.
-  const MaterialLocalizations();
+  /// Construct an object that defines the material widgets' localized strings
+  /// for the given `locale`.
+  const MaterialLocalizations(this.locale) : assert(locale != null);
 
   /// The locale for which the values of this class's localized resources
   /// have been translated.
-  Locale get locale => const Locale('en', 'US');
+  final Locale locale;
```
Comment (Hixie) [inline packages/flutter/lib/src/material/material_localizations.dart]: This class is also the interface that other MaterialLocalizations implementations will implement or extend, so it doesn't really make sense for this to be a field.

File: packages/flutter/lib/src/material/material_localizations.dart (hunk 1 above)
Comment (Hixie) [reply packages/flutter/lib/src/material/material_localizations.dart]: Specifically, I'd like this class to be something you can extend without fear that you'll be dragging in code for any languages you don't know about.

**Comment by Hixie** [comment]
Presumably you also need to provide WidgetsLocalizations for each locale to set the textDirection, no?

File: packages/flutter/lib/src/material/i18n/localizations.dart [hunk 3]
```diff
@@ -0,0 +1,80 @@
+// Copyright 2017 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+// This file has been automatically generated.  Please do not edit it manually.
+// To regenerate the file, use:
+// dart gen_localizations packages/flutter/lib/src/material/i18n material
+
+const Map<String, Map<String, String>> localizations = const <String, Map<String, String>> {
```
Comment (Hixie) [inline packages/flutter/lib/src/material/i18n/localizations.dart]: This is great!

File: packages/flutter/lib/src/material/material_localizations.dart (hunk 1 above)
Comment (HansMuller) [reply packages/flutter/lib/src/material/material_localizations.dart]: OK, I'll make this real interface, and add a class (DefaultMaterialLocalizations) that implements it.

File: packages/flutter/lib/src/material/material_localizations.dart [hunk 4]
```diff
@@ -48,3 +47,45 @@ class MaterialLocalizations {
...
+  const DefaultMaterialLocalizations(this.locale) : assert(locale != null);
+
+  /// The locale for which the values of this class's localized resources
+  /// have been translated.
+  final Locale locale;
+
+  String _lookup(String name) {
+    final Map<String, String> nameToValue = localizations[locale.toString()]
+      ?? localizations[locale.languageCode]
+      ?? localizations['en'];
```
Comment (Hixie) [inline packages/flutter/lib/src/material/material_localizations.dart]: we shouldn't need to do this here, the load method should just map from the given Locale to the nearest matching supported locale and return the object appropriately configured. That way we don't have to do this dance every build, for instance.

File: packages/flutter/lib/src/widgets/localizations.dart [hunk 5]
```diff
@@ -383,7 +391,6 @@ class _LocalizationsState extends State<Localizations> {
   T resourcesFor<T>(Type type) {
     assert(type != null);
     final T resources = _typeToResources[type];
-    assert(resources.runtimeType == type);
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/localizations.dart]: you want to assert that the type matches, though. Probably using [`TypeMatcher`](https://master-docs-flutter-io.firebaseapp.com/flutter/widgets/TypeMatcher-class.html).

**Comment by Hixie** [comment]
LGTM if you move the logic from _lookup to load().

**Comment by Hixie** [comment]
This patch introduced two undocumented members.

**Pull request merged**